crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,295 @@
+                                .map(|highest| highest.saturating_add(1))
+                                .unwrap_or(0),
+                        );
+
+                        // Schedule as many checkpoints as possible in order to catch up quickly
+                        for seq_num in next_to_exec..=last_to_exec {
+                            self.schedule_checkpoint(seq_num, &mut finished).await;
+
+                            if finished.len() >= task_limit || self.end_of_epoch {
+                                break;
+                            }
+                        }
+                    }
+                    // In this case, messages in the mailbox have been overwritten
+                    // as a result of lagging too far behind. We can simply continue the loop,
+                    // as the next call to recv() should return the overwritten checkpoint and we
+                    // will attempt to execute all checkpoints from (highest_executed_checkpoint + 1)
+                    // up to the result of the next recv() call.
+                    Err(RecvError::Lagged(num_skipped)) => {
+                        warn!(
+                            "Checkpoint Execution Recv channel overflowed {:?} messages",
+                            num_skipped,
+                        );
+                        self.metrics
+                            .checkpoint_exec_recv_channel_overflow
+                            .inc_by(num_skipped);
+                    }
+                    Err(RecvError::Closed) => {
+                        info!("Checkpoint Execution Sender (StateSync) closed channel");
+                        return None;
+                    }
+                },
+            }
+        }
+    }
+
+    async fn schedule_checkpoint(
+        &mut self,
+        seq_num: CheckpointSequenceNumber,
+        finished: &mut CheckpointExecutionBuffer,
+    ) {
+        let next_checkpoint = match self
+            .checkpoint_store
+            .get_checkpoint_by_sequence_number(seq_num)
+        {
+            Ok(Some(new_checkpoint)) => new_checkpoint,
+            Ok(None) => {
+                self.metrics.checkpoint_exec_errors.inc();
+                panic!(
+                    "Checkpoint {:?} does not exist in checkpoint store",
+                    seq_num,
+                );
+            }
+            Err(err) => {
+                self.metrics.checkpoint_exec_errors.inc();
+                panic!("Failed to fetch checkpoint {:?}: {:?}", seq_num, err);
+            }
+        };
+
+        // Since StateSync is guaranteed to produce checkpoint sync messages
+        // in monotonically increasing order, then a mismatch of epoch would
+        // mean that we failed to execute reconfig at the epoch boundary. This
+        // is an invalid state that should not happen under normal operation
+        assert_eq!(next_checkpoint.epoch(), self.authority_state.epoch());
```

Comment:
I am not sure how this could be ensured? In particular, what's the strategy to block wait for reconfig?